[FLINK-18226][runtime] Fix ActiveResourceManager requesting extra workers on termination of existing workers.

xintongsong opened the pull request:
## What is the purpose of the change

This PR fixes the problem that `ActiveResourceManager` may request unnecessary extra workers on termination of existing workers.

## Brief change log

The equation previously used by `ActiveResourceManager` for deciding whether and how many new workers to request on termination of existing workers is:
`workersToRequest = workersRequiredBySlotManager - workersRequestedByResourceManagerNotAllocated`

If there are workers allocated, being started but not yet registered, unnecessary new workers will be requested.

Now the equation is changed to: 
`workersToRequest = workersRequiredBySlotManager - workersRequestedByResourceManagerNotRegistered`

## Verifying this change

New test cases added in `KubernetesResourceManagerTest` and `YarnResourceManagerTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManagerGateway.java:
```diff
@@ -104,13 +104,15 @@
 	 * @return Future which is completed with {@link Acknowledge} once the slot report has been received.
 	 */
-	CompletableFuture<Acknowledge> sendSlotReport(
+	CompletableFuture<Acknowledge> sendInitialSlotReport(
```
I think this is not only about how it is used by the `TaskExecutor`, but also what this method does. The implementation of this method in `ResourceManager` actually registers the `TaskExecutor` at the `SlotManager`.

However, this renaming is not absolutely necessary. I'm also ok with reverting it if consensus is not reached on this, given what we really want to fix this problem before the next RC.

tillrohrmann replied:
Currently, `SlotManagerImpl.registerTaskManager` will call `SlotManagerImpl.reportSlotStatus` if the TM is already registered. Hence, the behaviour should be correct on the `RM` side.

xintongsong commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/KubernetesResourceManagerTest.java:
```diff
@@ -390,31 +429,7 @@ public void testPreviousAttemptPodAdded() throws Exception {
-	}
-
-	@Test
-	public void testDuplicatedPodAdded() throws Exception {
```
This test case was verifying that when duplicated `added` event is received for the same pod, only the first one decreases the pending counter.

I removed before realizing we need two different counters. When there's only one counter that decreases at registration of new worker, there won't be any behavior differences between handling the first and the duplicated `added` event.

This is no longer true since now we have two counters, one of which is still decreased on new worker `added`. I can add this test case back and make it verify agains the `requestedNotAllocatedWorkerCounter` counter. Although currently the `KubernetesResourceManager` is not relying on `requestedNotAllocatedWorkerCounter` (only `YarnResourceManager` is using it), it does not hurt to make sure the counters are properly maintained, incase `KubernetesResourceManager` uses it in the future.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -404,13 +404,17 @@ private void stopResourceManagerServices() throws Exception {
 
 		if (workerTypeWorkerRegistration.getInstanceID().equals(taskManagerRegistrationId)) {
-			slotManager.registerTaskManager(workerTypeWorkerRegistration, slotReport);
+			registerTaskManagerAtSlotManager(workerTypeWorkerRegistration, slotReport);
```
I think this is a good suggestion on how to providing sub-classes obstruct the registration. Thanks for that. 

What I don't understand is why do we need the `SlotManager.registerTaskManager` returns whether it is a new registration or a duplicated one.

I think the following changes should work, with a no-op implementation for `onTaskManagerRegistration` in the base class, overridden in sub classes. 
```suggestion
			slotManager.registerTaskManager(workerTypeWorkerRegistration, slotReport);
			onTaskManagerRegistration(workerTypeWorkerRegistration);
```

tillrohrmann replied:
I think we might call `onTaskManagerRegistration` multiple times for the same `TM` registration if `registerTaskManager` does not tell use whether it is a new registration or not.

xintongsong commented:
Thanks for the review, @tillrohrmann.
I've addressed your comments. Please take another look.

BTW. The revert commit is added to clearly show what have been changed since your last review. I'll remove it together with the original renaming commit at merging, if you're good with the changes.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/ResourceManager.java:
```diff
@@ -404,15 +404,16 @@ private void stopResourceManagerServices() throws Exception {
 		if (workerTypeWorkerRegistration.getInstanceID().equals(taskManagerRegistrationId)) {
-			registerTaskManagerAtSlotManager(workerTypeWorkerRegistration, slotReport);
+			slotManager.registerTaskManager(workerTypeWorkerRegistration, slotReport);
+			onTaskManagerRegistration(workerTypeWorkerRegistration);
```
I think we might call `onTaskManagerRegistration` multiple times if the `TM` decides to call `sendSlotReport` repeatedly.

xintongsong commented:
Thanks @tillrohrmann, your last comment makes sense to me.
I've updated the PR:
- Rebased onto the latest master branch
- Squashed previous commits
- Addressed your last comments in the new fixup commit.
